### PR TITLE
Add LongCat T2V (Base, Distillation and Refinement) Support to FastVideo

### PR DESCRIPTION
**Pull request by alexzms**
## Summary
This PR integrates LongCat-Video into FastVideo as a first-class text-to-video (T2V) pipeline, including:
- Native LongCat DiT model config and registration
- A LongCat pipeline that supports base 480p generation
- Distillation via LoRA
- 480p → 720p refinement (two-stage pipeline)
- Optional Block Sparse Attention (BSA) and 3D RoPE support
- Utilities for checkpoint conversion

### 1. Model & Pipeline Configuration
* `fastvideo/configs/models/dits/longcat.py`
  * Defines `LongCatVideoArchConfig` and `LongCatVideoConfig` for the native LongCat DiT.
  * Adds parameter name mappings to convert official LongCat weights to FastVideo naming (embedders, AdaLN, self/cross-attn, FFN, final layer).
  * Exposes BSA-related fields and video-specific settings (3D patches, caption channels, etc.).

* `fastvideo/configs/pipelines/longcat.py`
  * Adds `LongCatDiTArchConfig` for Phase 1 wrapper compatibility.
  * Adds `LongCatT2V480PConfig` (base 480p pipeline) and `LongCatT2V704PConfig` (704p refinement with BSA enabled).

* `fastvideo/configs/models/dits/__init__.py` / `fastvideo/models/registry.py` / `fastvideo/configs/pipelines/registry.py` / `fastvideo/pipelines/pipeline_registry.py`
  * Wires LongCat configs and models into the existing registry:
    * Registers LongCat DiT classes and `LongCatPipeline`.
    * Adds pipeline detection and fallback under the `"longcat"` key.

### 2. LongCat Pipeline & Stages (Base + Refinement)
* `fastvideo/pipelines/basic/longcat/longcat_pipeline.py`
  * Implements `LongCatPipeline` as a composed pipeline with LoRA support.
  * Assembles stages for:
    * text encoding, timestep prep, latent prep, denoising, decoding
    * LongCat-specific refine stages (`LongCatRefineInitStage`, `LongCatRefineTimestepStage`).
  * Enables runtime BSA configuration from pipeline config / CLI and propagates parameters to transformer blocks.

* `fastvideo/pipelines/stages/longcat_refine_init.py`
  * Initializes 480p → 720p refinement:
    * Loads stage1 video (path or in-memory frames).
    * Upsamples spatially/temporally, applies temporal padding compatible with VAE/BSA.
    * VAE-encodes, normalizes latents, and mixes with noise according to `t_thresh`.
    * Stores padding metadata in `batch` for later cropping.

* `fastvideo/pipelines/stages/longcat_refine_timestep.py`
  * Builds LongCat refinement timesteps starting at `t_thresh` and updates scheduler timesteps/sigmas accordingly.
* `fastvideo/pipelines/stages/longcat_denoising.py`
  * LongCat-specific denoising loop:
    * Batched CFG with CFG-zero optimal guidance scale.
    * Negates `noise_pred` to match flow-matching scheduler convention.

* `fastvideo/pipelines/pipeline_batch_info.py` / `fastvideo/configs/sample/base.py`
  * Extends `ForwardBatch` and `SamplingParam` with LongCat refine fields: `refine_from`, `t_thresh`, `spatial_refine_only`, `num_cond_frames`, `stage1_video`.
  * Adds corresponding CLI args (`--refine-from`, `--t-thresh`, etc.).

* `fastvideo/pipelines/stages/latent_preparation.py` / `fastvideo/pipelines/stages/decoding.py`
  * Adjusts latent scaling for pre-initialized latents in refine mode (no double `init_noise_sigma`).
  * Crops extra padded frames after decoding using refine padding metadata.

* `fastvideo/pipelines/stages/utils.py`
  * Adds aspect-ratio bucket tables and `get_bucket_config()` used to select resolutions for 480p/720p LongCat.

### 3. RoPE, LoRA & BSA Triton kernel support
* `fastvideo/layers/rotary_embedding_3d.py`
  * Adds 3D RoPE implementation for video transformers, splitting head dim into (T, H, W) components.

* `fastvideo/layers/lora/linear.py` / `fastvideo/pipelines/lora_pipeline.py` / `fastvideo/fastvideo_args.py`
  * Implements LoRA alpha scaling in merge logic (`alpha = lora_alpha / rank`).
  * Teaches `LoRAPipeline` to parse/store `*.lora_alpha` alongside `lora_A`, `lora_B`.
  * Introduces CLI flags for loading LoRA adapters (`--lora-path`, `--lora-nickname`, `--lora-target-modules`).

* `fastvideo/third_party/longcat_video/block_sparse_attention/*`
  * Integrates LongCat’s Triton-based Block Sparse Attention kernels and helper utilities (including p2p communication for context parallelism).

### 4. Checkpoint Conversion, Inference Scripts
* `scripts/checkpoint_conversion/longcat_to_fastvideo.py`
  * Converts official LongCat-Video checkpoints into a FastVideo-compatible layout using the new param name mappings.

* `scripts/inference/v1_inference_longcat*.sh`
  * Example scripts for:
    * 480p LongCat T2V generation
    * 480p distillation with LoRA
    * 480p → 720p refinement from an existing video (with BSA + refinement LoRA)

## Tests
- **Base LongCat T2V**: Ran generation on all prompts under `assets/` and verified successful 480p video outputs for each prompt.
- **Distill + Refine pipeline**: Used a representative prompt to test the full two-stage flow:
  1) 480p distilled generation, and  
  2) 480p → 720p refinement from the generated video.

**Known limitations:**
- Distill and refine are currently triggered by two separate scripts and require manual chaining. If needed in the future, we plan to add a patch that unifies both stages into a single end-to-end command.
- We observe that 2-GPU generation is currently slower than 1-GPU generation for the LongCat pipeline; this performance issue is under investigation.

### CODE REVIEW

File: fastvideo/configs/pipelines/longcat.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    text = text.strip()
+    
+    return text
+
+
+def umt5_postprocess_text(outputs: BaseEncoderOutput) -> torch.Tensor:
+    """Postprocess UMT5/T5 encoder outputs to fixed length 512 embeddings.
+
+    TODO: Not sure if using same procedure as t5 postprocess_text function is correct.
+    """
```
[the remark this answers is not in the document]
Comment (alexzms) [reply fastvideo/configs/pipelines/longcat.py:75]: This is solved. It should be save to remove the TODO comments.

File: fastvideo/configs/pipelines/longcat.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
+from collections.abc import Callable
```
Comment (SolitaryThinker) [inline fastvideo/configs/pipelines/longcat.py:3]: add license header `# SPDX-License-Identifier: Apache-2.0`

File: fastvideo/configs/pipelines/longcat.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    - VAE (8x): 88×160×96  
+    - Patch [1,2,2]: 44×80×96
+    - chunk [4,4,4]: 96%4=0, 44%4=0, 80%4=0 ✅
+    
+    This configuration matches the original LongCat refinement stage parameters.
+    """
+    
+    # Enable BSA by default for 704p
+    enable_bsa: bool = True
+
```
Comment (SolitaryThinker) [inline fastvideo/configs/pipelines/longcat.py]: add newline char at eof

File: fastvideo/pipelines/basic/longcat/longcat_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        # LongCat uses FlowMatchEulerDiscreteScheduler which is already loaded
+        # from the model_index.json, so no need to override
+        
+        # Enable BSA (Block Sparse Attention) if configured
+        pipeline_config = fastvideo_args.pipeline_config
+        transformer = self.get_module("transformer", None)
+        if transformer is None:
+            return
+        # If user toggles BSA via CLI/config
+        if getattr(pipeline_config, 'enable_bsa', False):
```
Comment (SolitaryThinker) [inline fastvideo/pipelines/basic/longcat/longcat_pipeline.py]: can we avoid using getattr, especially for structures such as pipeline_config that should have these fields. These can sometimes cause silent bugs that are hard to debug

File: fastvideo/pipelines/basic/longcat/longcat_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        self.add_stage(
+            stage_name="decoding_stage",
+            stage=DecodingStage(
+                vae=self.get_module("vae"),
+                pipeline=self
+            )
+        )
+
+
+EntryClass = LongCatPipeline
```
Comment (SolitaryThinker) [inline fastvideo/pipelines/basic/longcat/longcat_pipeline.py]: newline char

File: .gitignore [hunk 6]
```diff
@@ -30,6 +30,8 @@ env
 **/build/
 **.pyc
 **.txt
+add-longcat-video/
```
Comment (jzhang38) [inline .gitignore]: remove

File: .gitignore [hunk 7]
```diff
@@ -70,3 +72,4 @@ docs/distillation/examples/
 
 dmd_t2v_output/
 preprocess_output_text/
+weights/
```
Comment (jzhang38) [inline .gitignore]: remove

File: fastvideo/layers/rotary_embedding_3d.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        """
+        Precompute and register frequencies for a given grid size.
+        
+        Args:
+            grid_size: (T, H, W) tuple of grid dimensions
+        """
+        if grid_size not in self.freqs_dict:
+            self.freqs_dict[grid_size] = self.precompute_freqs_3d(grid_size)
+
+    def precompute_freqs_3d(self, grid_size: tuple[int, int, int]) -> torch.Tensor:
```
Comment (jzhang38) [inline fastvideo/layers/rotary_embedding_3d.py]: Is it the same with https://github.com/hao-ai-lab/FastVideo/blob/681f1583f9b27cb8e5d5f8dd02925bd307ee168b/fastvideo/layers/rotary_embedding.py#L423 ?

File: scripts/inference/v1_inference_longcat_refine_fromvideo.sh [hunk 9]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
```
Comment (jzhang38) [inline scripts/inference/v1_inference_longcat_refine_fromvideo.sh:1]: also add scripts under examples/inference?

File: fastvideo/pipelines/stages/utils.py [hunk 10]
```diff
@@ -77,3 +77,116 @@ def retrieve_timesteps(
             raise ValueError("scheduler.timesteps is None after set_timesteps")
         num_inference_steps = len(timesteps)
     return timesteps, num_inference_steps
+
```
Comment (jzhang38) [inline fastvideo/pipelines/stages/utils.py]: should those be put under longcat configs?

File: fastvideo/layers/lora/linear.py [hunk 11]
```diff
@@ -395,4 +395,4 @@ def replace_submodule(model: nn.Module, module_name: str,
     parent = model.get_submodule(".".join(module_name.split(".")[:-1]))
     target_name = module_name.split(".")[-1]
     setattr(parent, target_name, new_module)
-    return new_module
+    return new_module
```
Comment (SolitaryThinker) [inline fastvideo/layers/lora/linear.py]: add newline char

File: fastvideo/pipelines/basic/longcat/longcat_pipeline.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    """
+
+    _required_config_modules = [
+        "text_encoder", "tokenizer", "vae", "transformer", "scheduler"
+    ]
+
+    def initialize_pipeline(self, fastvideo_args: FastVideoArgs):
+        """Initialize LongCat-specific components."""
+        # LongCat uses FlowMatchEulerDiscreteScheduler which is already loaded
+        # from the model_index.json, so no need to override
```
Comment (SolitaryThinker) [inline fastvideo/pipelines/basic/longcat/longcat_pipeline.py]: can remove

File: fastvideo/pipelines/basic/longcat/longcat_pipeline.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+        # Enable BSA (Block Sparse Attention) if configured
+        pipeline_config = fastvideo_args.pipeline_config
+        transformer = self.get_module("transformer", None)
+        if transformer is None:
+            return
```
Comment (SolitaryThinker) [inline fastvideo/pipelines/basic/longcat/longcat_pipeline.py]: this should raise a Runtime error instead of silently returning

File: fastvideo/pipelines/basic/longcat/longcat_pipeline.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        if pipeline_config.enable_bsa:
+            # Build effective BSA params:
+            # 1) from explicit CLI overrides if provided
+            # 2) else from pipeline_config.bsa_params
+            # 3) else fall back to reasonable defaults
+            bsa_params_cfg = getattr(pipeline_config, 'bsa_params', None) or {}
+            sparsity = getattr(pipeline_config, 'bsa_sparsity', None)
+            cdf_threshold = getattr(pipeline_config, 'bsa_cdf_threshold', None)
+            chunk_q = getattr(pipeline_config, 'bsa_chunk_q', None)
+            chunk_k = getattr(pipeline_config, 'bsa_chunk_k', None)
```
Comment (SolitaryThinker) [inline fastvideo/pipelines/basic/longcat/longcat_pipeline.py]: if longcat's pipeline_config explicitly defines these args, then getattr should not be used here.

File: fastvideo/pipelines/stages/latent_preparation.py [hunk 15]
```diff
@@ -96,16 +96,24 @@ def forward(
...
+            )
+            if hasattr(self.scheduler, "init_noise_sigma"):
+                latents = latents * self.scheduler.init_noise_sigma
         else:
+            # Pre-initialized latents:
+            # - For LongCat refine (refine_from or stage1_video present), we should not re-scale by init_noise_sigma.
+            # - For other models, keep the original behavior.
             latents = latents.to(device)
+            if not (getattr(batch, "refine_from", None) is not None or getattr(
+                    batch, "stage1_video", None) is not None) and hasattr(
```
Comment (SolitaryThinker) [inline fastvideo/pipelines/stages/latent_preparation.py]: same as above for usage of getattr

File: fastvideo/pipelines/lora_pipeline.py [hunk 16]
```diff
@@ -191,19 +191,6 @@ def set_lora_adapter(self,
...
-                    # but store in lora_adapters with ".lora_alpha" suffix
-                    layer_name = name.replace(".lora_alpha", "")
-                    layer_name, _, _ = lora_param_names_mapping_fn(layer_name)
-                    target_name, _, _ = param_names_mapping_fn(layer_name)
-                    # Store alpha alongside weights with same target_name base
-                    alpha_key = target_name + ".lora_alpha"
-                    self.lora_adapters[lora_nickname][alpha_key] = weight.item(
-                    ) if weight.numel() == 1 else float(weight.mean())
-                    continue
-
```
Comment (SolitaryThinker) [inline fastvideo/pipelines/lora_pipeline.py:229]: just double checking that this code removal is intended?

**Review by SolitaryThinker — CHANGES_REQUESTED** [review]
Okay, overall I think we should avoid using getattr as much as possible, please take a pass over the PR and think carefully if getattr is the right way. 

otherwise I think this PR is ready to be merged. I know there's other inference features such as i2v, and those should be in a separate PR.

**Comment by alexzms** [comment]
Successfully implemented SSIM and verified it on the 4x L40S platform. Currently, the thresholds are set to 0.93 for distill/base and 0.90 for refine. Please let me know if these values align with expectations (or if they need tuning).

**Comment by alexzms** [comment]
The SSIM tests are still timing out (>60 mins) even with reduced steps. I suggest merging the Longcat t2v without ssim for now, and I'll handle the SSIM efficiency optimization in a separate PR.
